force-app/main/default/classes/DMLOperations.cls

```diff
@@ -146,9 +157,39 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+         // Query the Account Id based on the account name
+        Account acc = new Account();
+        List<Account> accounts = [SELECT Id, Name FROM Account WHERE Name = :accountName];
+        if (accounts.size() == 0) {
+            acc.Name = accountName;
+            insert acc;
+        } else {
+            acc = accounts.get(0);
+        }
+        // List to hold Opportunity objects     
+        List<Opportunity> opportunities = [SELECT Id, Name FROM Opportunity WHERE AccountId = :acc.Id];  
+        // Create an Opportunity with the given name
+        if (opportunities.size() == 0) {
+            for (String oppName : oppNames) {
+                Opportunity opp = new Opportunity(Name = oppName, AccountId = acc.Id);
+                opp.StageName = 'Qualification';
+                opp.CloseDate = Date.today().addMonths(3);
+                opportunities.add(opp);
+            }
+        } else {
+        // Add the Opportunity to the list
+        List<Opportunity> opportunitiesToUpsert = new List<Opportunity>();
+            // Loop through each Opportunity name
+            if (opportunities.size() == oppNames.size()) {
+                for (Opportunity opp : opportunities) {
+                    if (!oppNames.contains(opp.Name)) {
+                        Integer i = opportunities.indexOf(opp);
+                        opp.Name = oppNames.get(i);
+                    }
+                }
+            }
```

Comment:
Using a map, it's possible to avoid a lot of these for loops. What if you map the Name of the opportunity to its Id or SObject and see if you can use that to use the given opportunityNames somehow to detect duplicates, or take advantage of the upsert functionality.